pkg/csi/service/wcpguest/controller.go

```diff
@@ -772,29 +780,40 @@ func controllerUnpublishForBlockVolume(ctx context.Context, req *csi.ControllerU
 	for !isVolumeDetached {
 		log.Debugf("Waiting for update on VirtualMachine: %q", virtualMachine.Name)
 		// Block on update events
-		event := <-watchVirtualMachine.ResultChan()
-		vm, ok := event.Object.(*vmoperatortypes.VirtualMachine)
+		event, ok := <-watchVirtualMachine.ResultChan()
 		if !ok {
 			msg := fmt.Sprintf("Watch on virtualmachine %q timed out", virtualMachine.Name)
 			log.Error(msg)
 			return nil, status.Errorf(codes.Internal, msg)
 		}
-		if vm.Name != virtualMachine.Name {
-			log.Debugf("Observed vm name: %q, expecting vm name: %q, volumeID: %q. Continuing...", vm.Name, virtualMachine.Name, req.VolumeId)
-			continue
-		}
-		isVolumeDetached = true
-		for _, volume := range vm.Status.Volumes {
-			if volume.Name == req.VolumeId {
-				log.Debugf("Volume %q still exists in VirtualMachine %q status", volume.Name, virtualMachine.Name)
-				isVolumeDetached = false
-				if volume.Attached && volume.Error != "" {
-					msg := fmt.Sprintf("failed to detach volume %q from VirtualMachine %q with Error: %v", volume.Name, virtualMachine.Name, volume.Error)
-					log.Error(msg)
-					return nil, status.Errorf(codes.Internal, msg)
+		switch event.Type {
+		case watch.Added, watch.Modified:
+			vm, ok := event.Object.(*vmoperatortypes.VirtualMachine)
+			if !ok {
+				msg := fmt.Sprintf("Watch on virtualmachine %q timed out", virtualMachine.Name)
+				log.Error(msg)
+				return nil, status.Errorf(codes.Internal, msg)
+			}
+			if vm.Name != virtualMachine.Name {
+				log.Debugf("Observed vm name: %q, expecting vm name: %q, volumeID: %q. Continuing...", vm.Name, virtualMachine.Name, req.VolumeId)
+				continue
+			}
+			isVolumeDetached = true
+			for _, volume := range vm.Status.Volumes {
+				if volume.Name == req.VolumeId {
+					log.Debugf("Volume %q still exists in VirtualMachine %q status", volume.Name, virtualMachine.Name)
+					isVolumeDetached = false
+					if volume.Attached && volume.Error != "" {
+						msg := fmt.Sprintf("failed to detach volume %q from VirtualMachine %q with Error: %v", volume.Name, virtualMachine.Name, volume.Error)
+						log.Error(msg)
+						return nil, status.Errorf(codes.Internal, msg)
+					}
+					break
 				}
-				break
 			}
+		case watch.Deleted:
+			log.Infof("VirtualMachine %s/%s deleted. Assuming volume %s was detached.", c.supervisorNamespace, req.NodeId, req.VolumeId)
+			return &csi.ControllerUnpublishVolumeResponse{}, nil
```

Comment:
You may get delete event on a completely different VM as well. I dont think we should assume that the volume is detached in such a case. Can you check whether the vm.Name is same as virtualMachine.Name in this case block as well?

Reply:
done!